atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/ObjectPersister.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts.bucketingthings;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.google.errorprone.annotations.CompileTimeConstant;
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeUncheckedIoException;
+import java.io.IOException;
+
+// Without a type T, we can't safely implement tryDeserialize without an error of unsafe type cases.
+// Having T also helps give some type safety to prevent accidental screwups when using the many reader writers.
+final class ObjectPersister<T> {
+    private final ObjectReader reader;
+    private final ObjectWriter writer;
+    private final Class<T> clazz; // Partly exists to "consume" T, but also so that our logs can be more specific.
+    private final LogSafety logSafety;
+
+    private ObjectPersister(ObjectReader reader, ObjectWriter writer, Class<T> clazz, LogSafety logSafety) {
+        this.reader = reader;
+        this.writer = writer;
+        this.clazz = clazz;
+        this.logSafety = logSafety;
+    }
+
+    static <T> ObjectPersister<T> of(ObjectMapper objectMapper, Class<T> clazz, LogSafety logSafetyOfValues) {
+        return new ObjectPersister<>(
+                objectMapper.readerFor(clazz), objectMapper.writerFor(clazz), clazz, logSafetyOfValues);
+    }
+
+    public T tryDeserialize(byte[] bytes) {
+        try {
+            return reader.readValue(bytes);
+        } catch (IOException e) {
+            throw new SafeUncheckedIoException(
+                    "Failed to deserialise {} into {}", e, toArg("bytes", bytes), SafeArg.of("class", clazz.getName()));
+        }
+    }
+
+    public byte[] trySerialize(T object) {
+        try {
+            return writer.writeValueAsBytes(object);
+        } catch (JsonProcessingException e) {
+            throw new SafeUncheckedIoException(
+                    "Failed to serialise {} from {}", e, toArg("object", object), SafeArg.of("class", clazz.getName()));
+        }
+    }
+
+    private <K> Arg<K> toArg(@CompileTimeConstant String name, K value) {
```

Comment:
you'll need to block `SafeLoggingPropagation` on this probably, since it'll think this is unsafe

Reply:
Yeah - my bad, as noted internally, hadn't got round to the self-review due to other immediate priorities.